module/zfs/dmu_recv.c

```diff
@@ -2590,9 +2668,22 @@ receive_writer_thread(void *arg)
 			kmem_free(rrd->payload, rrd->payload_size);
 			rrd->payload = NULL;
 		}
-		kmem_free(rrd, sizeof (*rrd));
+		/*
+		 * EAGAIN indicates that this record has been saved (on
+		 * raw->write_batch), and will be used again, so we don't
+		 * free it.
+		 */
+		if (err != EAGAIN) {
```

Comment:
Could we assert that if err == EAGAIN, rrd->header.drr_type == DRR_WRITE?

Reply:
That would be correct, but I don't think that this code depends on it.  It would seem straightforward to batch together other types of records, in which case we'd want to not free them here either (e.g. DRR_OBJECT)

Reply:
Mostly I was thinking here that it might be beneficial to verify that nothing else is returning EAGAIN for some other reason (and that they won't do so in the future without considering this logic) to avoid any possible leaks. Not a huge deal either way, though.